### PR TITLE
fix execution id creation & submission race

### PR DESCRIPTION
**Pull request by danielnorberg**
Create the execution id in the prepare state and add it to the run state as part of the submit event. This eliminates the race between k8s submission and execution id persistence that caused the runner to not be able to look up run states for newly started pods.

This also means that styx will not "lose" any pods that were in the middle of being created during a styx restart.

For backwards compatibility, the `executionId` field is kept in the `submitted` event to ensure that we can replay old events and do not lose execution id's when deploying this change.

### CODE REVIEW

**Comment by danielnorberg** [comment]
@ulzha @TinaRanic @bergman @fabriziodemaria @rouzwawi @honnix PTAL

**Comment by danielnorberg** [comment]
Realized that an issue with this PR is that if `started` is received before `submitted`, then `RunState` does not increment `tries`.

Perhaps we should emit `submitted` _before_ starting the container.

**Comment by honnix** [comment]
It is a a problem. It feels weird to have `submitted` before actual submission though.

**Comment by danielnorberg** [comment]
@honnix even worse, receiving `submitted` after `started` would throw `illegalTransition`.

**Comment by danielnorberg** [comment]
* Currently, receiving `submitted` throws `illegalTransition` unless the current state is `SUBMITTING`.
* The intent seems to be for the `RunState` state machine to receive `submitted` before events from the container like `started`.
* Currently, receiving `started` before `submitted` is prevented by discarding pod events.
* However, it seems that purpose of the watcher is to receive pod events and emit `started` when pods are created. Thus it seems undesirable to discard events for newly created pods just due to a race condition.

By emitting `submitted` before creating the pod we can preserve the `submitted` before `started` ordering. The `submitted` event would thus mean `pod creation has been asynchronously requested`: https://github.com/spotify/styx/pull/200

File: styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java [hunk 1]
```diff
@@ -71,18 +70,29 @@ public void transitionInto(RunState state) {
         // Perform rate limited submission on a separate thread pool to avoid blocking the caller.
         executor.submit(() -> {
           rateLimiter.acquire();
+
+          final RunSpec runSpec;
           try {
-            final String executionId = dockerRunnerStart(state);
-            // this is racy
```
Comment (ulzha) [inline styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java]: :star:

File: styx-common/src/main/java/com/spotify/styx/model/EventVisitor.java [hunk 2]
```diff
@@ -37,8 +38,9 @@
   R triggerExecution(@Getter WorkflowInstance workflowInstance, Trigger trigger);
   R info(@Getter WorkflowInstance workflowInstance, Message message);
   R dequeue(@Getter WorkflowInstance workflowInstance);
-  R submit(@Getter WorkflowInstance workflowInstance, ExecutionDescription executionDescription);
-  R submitted(@Getter WorkflowInstance workflowInstance, String executionId);
+  R submit(@Getter WorkflowInstance workflowInstance, ExecutionDescription executionDescription,
+      @Nullable String executionId);
```
Comment (ulzha) [inline styx-common/src/main/java/com/spotify/styx/model/EventVisitor.java]: In `RunSpec` we add `executionId`, yet in `ExecutionDescription` we don't add it but carry it alongside like this... Intended?

I think `ExecutionDescription` should eventually converge with `RunSpec` to be one value class... Any reason not to? Javadocs missing for both.

File: styx-common/src/main/java/com/spotify/styx/model/data/WFIExecutionBuilder.java [hunk 3]
```diff
@@ -117,17 +117,21 @@ public Void created(WorkflowInstance workflowInstance, String executionId, Strin
...
+      currExecutionId = executionId;
 
       return null;
     }
 
     @Override
     public Void submitted(WorkflowInstance workflowInstance, String executionId) {
       currWorkflowInstance = workflowInstance;
-      currExecutionId = executionId;
+      if (currExecutionId == null) {
```
Comment (ulzha) [inline styx-common/src/main/java/com/spotify/styx/model/data/WFIExecutionBuilder.java]: What's the check for? Should this be an assertion that we don't get changing execution ids, and be logged as error?

File: styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java [hunk 4]
```diff
@@ -378,9 +374,12 @@ private synchronized void tryPollPods() {
       if (!workflowInstance.isPresent()) {
         continue;
       }
-      lookupPodRunState(pod, workflowInstance.get(), () -> cleanup(workflowInstance.get(),
-                                                                   pod.getMetadata().getName()))
-          .ifPresent(runState -> emitPodEvents(Watcher.Action.MODIFIED, pod, runState));
+      final Optional<RunState> runState = lookupPodRunState(pod, workflowInstance.get());
+      if (runState.isPresent()) {
```
Comment (ulzha) [inline styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java]: Nice.

File: styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java [hunk 5]
```diff
@@ -71,18 +70,29 @@ public void transitionInto(RunState state) {
...
             LOG.error("Unable to start docker procedure.", e);
             stateManager.receiveIgnoreClosed(Event.halt(state.workflowInstance()));
+            return;
+          }
+
+          // Emit submitted event first to guarantee it is observed before events from the pod
+          final Event submitted = Event.submitted(state.workflowInstance(), runSpec.executionId());
+          try {
+            stateManager.receive(submitted);
+          } catch (StateManager.IsClosed isClosed) {
```
Comment (ulzha) [inline styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java]: Kind of unsure why we need to check these. `StateManager` closing means we're exiting, right? Not sending events anymore then is normal operation. It would be fine to exit the event handler thread right there with an unchecked exception.

File: styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java [hunk 6]
```diff
@@ -71,18 +70,29 @@ public void transitionInto(RunState state) {
...
+            LOG.warn("Could not emit 'submitted' event", isClosed);
+            return;
+          }
+
+          try {
+            LOG.info("running:{} image:{} args:{} termination_logging:{}", state.workflowInstance().toKey(),
```
Comment (ulzha) [inline styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java]: Add `toString` to `runSpec`?

File: styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java [hunk 7]
```diff
@@ -108,30 +118,29 @@ public void transitionInto(RunState state) {
...
+  private RunSpec createRunSpec(RunState state) throws ResourceNotFoundException {
     final Optional<ExecutionDescription> executionDescriptionOpt = state.data().executionDescription();
 
     final ExecutionDescription executionDescription = executionDescriptionOpt.orElseThrow(
         () -> new ResourceNotFoundException("Missing execution description for "
                                             + state.workflowInstance().toKey()));
 
+    final String executionId = state.data().executionId()
+        // Backwards compatibility fallback
+        .orElseGet(ExecutionDescriptionHandler::createExecutionId);
```
Comment (ulzha) [inline styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java]: Hard to think what will happen here... As we replay, we'll work with this real-looking unrelated ID?

File: styx-scheduler-service/src/test/java/com/spotify/styx/StyxSchedulerServiceFixture.java [hunk 8]
```diff
@@ -79,7 +81,7 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(StyxSchedulerServiceFixture.class);
 
-  static final String TEST_EXECUTION_ID = "execution_1";
+//  static final String TEST_EXECUTION_ID = "execution_1";
```
Comment (ulzha) [inline styx-scheduler-service/src/test/java/com/spotify/styx/StyxSchedulerServiceFixture.java]: Commented out code.

File: styx-scheduler-service/src/test/java/com/spotify/styx/state/handlers/DockerRunnerHandlerTest.java [hunk 9]
```diff
@@ -190,7 +195,7 @@ public void shouldFailIfDockerRunnerRaisesException() throws Exception {
     stateManager.initialize(runState);
     dockerRunnerHandler.transitionInto(runState);
 
-    verify(stateManager, timeout(60_000)).receive(any(Event.class));
+    verify(stateManager, timeout(60_000).times(2)).receive(any(Event.class));
```
Comment (ulzha) [inline styx-scheduler-service/src/test/java/com/spotify/styx/state/handlers/DockerRunnerHandlerTest.java]: Where does 2 come from?

**Comment by ulzha** [comment]
Also, please document the updated semantics of the `submitted` event. (At least the SUBMITTING state description in design-overview.md seems wrong now.)

File: styx-common/src/main/java/com/spotify/styx/model/EventVisitor.java (hunk 2 above)
Comment (danielnorberg) [inline styx-common/src/main/java/com/spotify/styx/model/EventVisitor.java]: Both `executionId` and `executionDescription` are (already) fields in `RunState.data()`. Adding `executionId` to `ExecutionDescription` would be redundant. Moving `executionId` from `RunState.data()` to `ExecutionDescription` could be an option, but not sure we would want that. Either way it would be a bigger and backwards incompatible change.

`RunSpec` is not part of `RunState`. It is only a part of the protocol between `DockerRunner` and `DockerRunnerHandler`.

Although `ExecutionDescription` and `RunSpec` are almost identical, I think they might live in separate bounded contexts and that DRY might not apply here. E.g. we would not want to add `trigger` to `ExecutionDescription` just because `DockerRunner` needs it and it is currently in `RunSpec`. Anyway, this might be a discussion worth having outside this PR.

File: styx-common/src/main/java/com/spotify/styx/model/data/WFIExecutionBuilder.java (hunk 3 above)
Comment (danielnorberg) [inline styx-common/src/main/java/com/spotify/styx/model/data/WFIExecutionBuilder.java]: I now believe this code to be incorrect. To be backwards compatible I think we need to keep unconditionally assigning `executionId` to `currExecutionId`.

Comment (danielnorberg) [inline styx-common/src/main/java/com/spotify/styx/model/data/WFIExecutionBuilder.java]: Changing execution id's is actually expected in some circumstances. E.g. a wfi that is retrying.

File: styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java (hunk 5 above)
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java]: Yeah, these checks have been bothering me as well. Actually, I think Styx should be crash-only and implemented in such a way that sudden restart does not disturb workflow execution. I.e. idempotent at-least-once event transition handling etc.

File: styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java (hunk 6 above)
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java]: `RunSpec` already has a `toString` as it is an `@AutoValue`. This is not new code though, just factored out from `dockerRunnerStart`, so let's do it in a separate PR.

File: styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java (hunk 7 above)
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java]: The intent here is to handle existing active WFIs that have already transitioned through `PREPARE`. As they transition through `SUBMITTING`, they will have no `executionId` in their `RunState.data()` and it will be created here by `DockerRunnerHandler` and added to `RunState.data()` through the `submitted` event. This is identical to current behavior.

New WFIs that transition through `PREPARE` will already have an `executionId` in their `RunState.data()` and that will be used.

Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java]: As for pure replays, they just go through the transitions defined by `RunState`. Old WFIs will have `executionId`  added to their state by a `submitted` event, new WFIs will have it added by a `submit` event.

File: styx-scheduler-service/src/test/java/com/spotify/styx/StyxSchedulerServiceFixture.java (hunk 8 above)
Comment (danielnorberg) [inline styx-scheduler-service/src/test/java/com/spotify/styx/StyxSchedulerServiceFixture.java]: oops 👍

File: styx-scheduler-service/src/test/java/com/spotify/styx/state/handlers/DockerRunnerHandlerTest.java (hunk 9 above)
Comment (danielnorberg) [inline styx-scheduler-service/src/test/java/com/spotify/styx/state/handlers/DockerRunnerHandlerTest.java]: The state manager receives two events:

1. `submitted`
2. `runError`

File: styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java (hunk 1 above)
Comment (danielnorberg) [inline styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/DockerRunnerHandler.java]: 🎆

**Comment by ulzha** [comment]
:+1:
